packages/xpath/src/lib/geo/Geotrace.ts

```diff
@@ -0,0 +1,82 @@
+import type { Geopoint } from './Geopoint.ts';
+import { geopointCodec, type GeopointRuntimeValue } from './geopointCodec.ts';
+import { GeotraceLine } from './GeotraceLine.ts';
+
+export type GeotracePoints = readonly [Geopoint, Geopoint, ...Geopoint[]];
+
+const isGeotracePoints = (
+	geopoints: readonly GeopointRuntimeValue[]
+): geopoints is GeotracePoints => {
+	return (
+		geopoints.length >= 2 &&
+		geopoints.every((geopoint) => {
+			return geopoint != null;
+		})
+	);
+};
+
+const collectLines = (geopoints: GeotracePoints): readonly GeotraceLine[] => {
+	return geopoints.reduce((acc, geopoint, i) => {
+		if (i === 0) {
+			return acc;
+		}
+
+		// Non-null assertion safe: we ensure at least 2 points, and skip index 0.
+		const start = geopoints[i - 1]!;
+		const end = geopoint;
+
+		acc.push(
+			new GeotraceLine({
+				start,
+				end,
+			})
+		);
+
+		return acc;
+	}, Array<GeotraceLine>());
+};
+
+export class Geotrace {
+	static fromEncodedGeotrace(encoded: string): Geotrace | null {
+		const geopoints = encoded
+			.replace(/\s*;\s*$/, '')
+			.split(/\s*;\s*/)
+			.map((value) => {
+				return geopointCodec.decodeValue(value);
+			});
+
+		return this.fromGeopoints(geopoints);
+	}
+
+	static fromEncodedValues(values: readonly string[]): Geotrace | null {
+		const [head, ...tail] = values;
+
+		if (head == null) {
```

Comment:
Question:

[In the docs](https://getodk.github.io/xforms-spec/#data-types), it says, `where the last geopoint’s latitude and longitude is not equal to the first`

Is that a validation that can be done here or better at the engine level? I'm thinking at engine level so it throws an exception, but I thought of double-checking

Reply:
OK. Even if the last and first are equal and there are no more points in between (so the array length is 2), it may be a valid value of something that didn't move 🤔

Reply:
Yeah. And as for how it affects XPath behavior, both `area` and `distance` will return `0` for two equal points.